### PR TITLE
PF-469 Add initial controlled notebook instance resource creation flight support

### PR DESCRIPTION
**Pull request by wchamber**
Add the initial steps and support for creating a controlled AI Notebook Instance with the Controlled Resource Service.

- Adds CRL google-notebooks dependency.
- Updates Stairway version to get newest FlightDebugInfo fields for testing.
- Adds connected tests for creating controlled AI notebook instance with some Step idempotency checking. This is unfortunately slow as creating an AI notebook instance takes ~3-7 minutes.
- Adds an initial OpenApi model definition for the AI notebook instance creation parameters.

Points of interest:
This follows the Bucket creation in storing the raw API creation parameters in the FlightMap. This locks us into changing the API model in ways supported by Stairway serialization. An alternative would be to make a Java POJO for JSON serialization. Or we can wait for serialization to shake out.

Right now the notebook instance network is hard coded. We could either expose the network as a part of the WM API, or retrieve and use any network/subnetwork in the appropriate GCE region. Probably a follow up PR.

TODOs for follow up PRs
- Setting up permissions for AI Notebook Instances and the service account on it
- Not hardcoding the network/subnetwork to use on the project
- More instance creation parameters.
- Add the actual OpenAPI creation endpoint.

### CODE REVIEW

**Comment by wchamber** [comment]
@rtitle FYI: First chunk of AI Notebook Instance as a controlled resource in WM.

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateControlledResourceFlight.java
Comment (zloery) [inline src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateControlledResourceFlight.java:52]: may not be needed with Tova's work on Sam exceptions

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateControlledResourceFlight.java
Comment (zloery) [inline src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateControlledResourceFlight.java:33]: is this a long enough maximum timeout? from the PR description I'd expect notebook description to take O(minutes)

**Review by zloery — COMMENTED** [review]
generally LGTM, some comments

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (ddietterich) [inline src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:157]: What if we have already created an instance of this name in this workspace?
Do we need another step that looks in the workspace metadata to make sure that is not the case?

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (ddietterich) [inline src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:70]: How can you tell the difference between the instance already existing from this flight and the instance existing from a previous flight creating it?

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (ddietterich) [inline src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:134]: From the comments in the open_api, I got the idea the vm image and the container image were mutually exclusive. Is that true? This could would allow both or neither to be set.

**Review by ddietterich — COMMENTED** [review]
Looks like there might be a duplicate instance issue.

File: src/main/resources/api/service_openapi.yaml
Comment (wchamber) [reply src/main/resources/api/service_openapi.yaml]: 63 AFAICT. `(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)` 1 + max 61 + 1.

File: src/main/resources/api/service_openapi.yaml
Comment (wchamber) [reply src/main/resources/api/service_openapi.yaml]: The service account on the VM. Which in this case has no special privileges, so we'll have to consider how to handle this when we have a use case.
Added a comment.

File: src/main/java/bio/terra/workspace/service/resource/controlled/ControlledResourceService.java
Comment (wchamber) [reply src/main/java/bio/terra/workspace/service/resource/controlled/ControlledResourceService.java:126]: I wasn't even thinking it was necessary for the Service. The [WorkspaceApiController](https://github.com/DataBiosphere/terra-workspace-manager/blob/a1fd416f69c3c032abb33060c56753fbb88db283/src/main/java/bio/terra/workspace/app/controller/WorkspaceApiController.java#L419) and the [ControlledGcpResourceApiController](https://github.com/DataBiosphere/terra-workspace-manager/blob/a1fd416f69c3c032abb33060c56753fbb88db283/src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java#L135) use the job service directly to get their job results. If it's preferable to wrap the job service lookup in `ControlledResourceService` I'd like to do that in a follow up PR once I get to actually having endpoints for the controller.

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (wchamber) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:70]: I was going to rely on Jay's PR #291 to do a transaction on the ResourceDao to ensure uniqueness of the underlying cloud resource. As-is, it's broken as it will assume success.

I'll add a TODO to add a test to verify that multiple creations for the same instance fail once Jay's PR is merged.

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (wchamber) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:72]: AFAICT it's guaranteed to be on the initial create call. I can't find any documentation stating either way.
But if it can show up on the poll, we have an issue.

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (wchamber) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:83]: Yes. I don't know if there's a preferred style here. Undo's can't throw a RetryException, so maybe it's better to always return a StepResult?
This doesn't have an explicit exception, so without throwing a RetryException I wasn't sure how to encode any message for Stairway debugging what happened. A `StepResult` doesn't have its own error message field, other than an `Exception`. Rather than invent an exception, I went with throwing a new RetryException with my own message.
Happy to do something else here.

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (wchamber) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:95]: pure (tiny) style preference. I'm in the camp that `static` is an easy signal to the reader that this is a pure method that doesn't mess with or rely on instance state. [go/java-practices/static#already_static](go/java-practices/static#already_static)

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (wchamber) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:134]: Yes. The GCP API will throw exceptions if both or neither are set. I was planning on doing creation parameter validation in the controller in a follow up PR.

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (wchamber) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:157]: per other comment, I was relying on Jay's PR #291 to do the resource uniqueness check at ResourceDao insertion time (the earlier `StoreMetadataStep` in this flight). I agree that without that this is a problem. I was not planning on blocking this PR on that, but if that doesn't get merged we'll have to come up with another solution here.
This is a generic problem for controlled resources and their underlying cloud resource.

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (zloery) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:150]: ah, I was comparing to BQ Datasets where you can specify location at the region level - e.g. `us-central1` is a valid location to create a Dataset. If GCP requires you specify a zone (e.g. `us-central1-a`) for a notebook instance, this is fine.

nit: can you rename `zone` to `region` here? after digging into this more it looks like `zone` is actually the more specific area like `us-central1-a` while `region` is the broader area `us-central1`: https://cloud.google.com/compute/docs/regions-zones. I don't know how consistent that is throughout GCP documentation, though

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateControlledResourceFlight.java
Comment (zloery) [inline src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateControlledResourceFlight.java:30]: RetryRule objects are stateful, I think you need a different object for each step here. Otherwise, you're sharing 10 total retries across both steps.

(We've definitely mixed this up elsewhere, PF-557 for other ones)

**Review by zloery — APPROVED** [review]
LGTM, acknowledging the concurrency issue that (I think) will be fixed by a DB check in Jay's PR.

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (ddietterich) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:83]: No need to change it. It just struck me odd looking at it. I don't think I've seen a case before where requesting Stairway retry is not in response to an exception. If this turns up more frequently, it might be worth adding a StepResult that takes a String instead of Exception

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (ddietterich) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:95]: OK, I'll keep that in mind in my code.

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (ddietterich) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:157]: OK, that sounds fine to me

**Review by ddietterich — APPROVED** [review]
LGTM

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (wchamber) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:150]: thanks, s/zone/region done.

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java
Comment (wchamber) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/notebook/CreateAiNotebookInstanceStep.java:72]: I don't think so. I think that's code that we're very unlikely to need. If our create AI notebook flights are failing, we should debugging why that's happening, not trying to predict what the issue is.
It also seems like it's not the style for the GCP operations API to use the operations.get endpoint HTTP status code for the result of the logical operation. The `Operation` response has an `error` field for describing errors about the operation. 
Reminiscent of our JobReport HTTP status code conversations.
https://cloud.google.com/ai-platform/notebooks/docs/reference/rest/v1/projects.locations.operations
Here, we're throwing the RetryException with the error retrieved from the operation response body.

File: src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateControlledResourceFlight.java
Comment (wchamber) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateControlledResourceFlight.java:30]: I did not know that. For my 2c, PF-557 SG. I don't know when I'd want to share a retry rule, so changing the `addStep` api to take a factory for retry rules to prevent that from happening seems less error prone.

Comment (zloery) [reply src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateControlledResourceFlight.java:30]: whoops, Dan's right here. I didn't see that stairway re-initialized RetryRules for each step

**Pull request merged**
